### PR TITLE
Offer changes and Category implementation

### PR DESCRIPTION
**Pull request by juan17rodriguez**

### CODE REVIEW

File: src/server/controllers/categories.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        const bodyResponse = { categories: categories.map((c) => Category.toDTO(c)) }
+        response.status(200).json(bodyResponse)
+    } catch (error) {
+        handleError(error, response)
+    }
+}
+
+export async function createAsync(request: Request, response: Response) {
+    try {
+        const { name, deleted } = request.body
```
Comment (CamiloNicolasMartinez) [inline src/server/controllers/categories.ts]: Deleted should not be received. It should be always setted to false when creating a category

File: src/server/interfaces/category.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        throw error
+    }
+}
+
+export async function createAsync(body: CategoryAttributes): Promise<Category | null> {
+    try {
+        const categoryInstance = await Category.createAsync(body)
+        if (!categoryInstance) { throw new ErrorPayload(500, 'Failed to create offer') }
+
+        return getAsync(categoryInstance.id)
```
Comment (CamiloNicolasMartinez) [inline src/server/interfaces/category.ts]: Why are we getting the just created category when we already have it on memory? That was OK when we were making the offers because it has some dependent entities that would be loaded this way. In this case its useless.
Just return the `Category.createAsync(body)`

File: src/server/interfaces/offer.ts [hunk 3]
```diff
@@ -1,8 +1,9 @@
-import { Offer, OfferAttributes } from '../models'
+import { Offer, OfferAttributes, OfferCategory, OfferCategoryAttributes } from '../models'
 import { validateStatus, validatePaymentMethod } from './helpers'
 import { logger } from '../services'
-import { ErrorPayload } from '../errorPayload';
-import { OfferPrice } from '../models/sequelize/offerPrice';
+import { ErrorPayload } from '../errorPayload'
+import { OfferPrice } from '../models/sequelize/offerPrice'
```
Comment (CamiloNicolasMartinez) [inline src/server/interfaces/offer.ts]: The model `OfferPrice` should be exported along the other models

**Review by ismaelpuricelli — APPROVED** [review]
Good work!

File: test/conf [hunk 4]
```diff
@@ -8,7 +8,7 @@ REDIS_PASSWORD=dontneed
 REDIS_USERNAME=dontneed
 
 DB_HOST = localhost
-DB_NAME_TEST = coopifyTest
-DB_USER = postgres
-DB_PASSWORD = dontneed
+DB_NAME_TEST = test_coopifydb
```
Comment (CamiloNicolasMartinez) [inline test/conf]: Undo this changes please

File: test/server/controllers/categories.test.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+import * as supertest from 'supertest'
+import * as _ from 'lodash'
+import { logInUser } from './helpers'
+import { UserAttributes } from '../../../src/server/models'
+import { app } from '../../../src/server'
+import { factory, createOffer, createCategory } from '../factory'
+import { logger } from '../../../src/server/services'
+
+const request = supertest(app)
+const createUser: UserAttributes = {
```
Comment (CamiloNicolasMartinez) [inline test/server/controllers/categories.test.ts]: We should use the createUser from the factory, or consider moving it to the factory

File: test/server/controllers/categories.test.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        context('Category already created', async () => {
+            let createCategoryClone, categoryId
+            beforeEach(async () => {
+                createCategoryClone = _.cloneDeep(createCategory)
+                const cat = await factory.create('category', createCategoryClone)
+                categoryId = cat.id
+            })
+            it('Should get the category list with the required parameters', async () => {
+                const res = await request.get(`/api/categories/${categoryId}`).expect(200)
+                logger.info('Response => ' + JSON.stringify(res.body))
```
Comment (CamiloNicolasMartinez) [inline test/server/controllers/categories.test.ts]: Remove this log

File: test/server/controllers/offers.test.ts [hunk 7]
```diff
@@ -60,27 +60,42 @@ describe('Offer Tests', async () => {
...
-            let createOfferClone, user
-            beforeEach(async () =>  {
-                const user = await factory.create('user', createUser) 
+            let createOfferClone, user, createCategoryClone, categoryId
+            beforeEach(async () => {
+                user = await factory.create('user', createUser)
                 createOfferClone = _.cloneDeep(createOffer)
                 createOfferClone.userId = user.id
+                createCategoryClone = _.cloneDeep(createCategory)
+                createCategoryClone.id = '319497a7-73fe-4e1c-b224-8f71d0ee0bb8'
```
Comment (CamiloNicolasMartinez) [inline test/server/controllers/offers.test.ts]: We should avoid getting randmon uuids hardcoded on our test code. We should use a uuid generating lib

File: test/server/controllers/offers.test.ts [hunk 8]
```diff
@@ -60,27 +60,42 @@ describe('Offer Tests', async () => {
...
+                const cat = await factory.create('category', createCategoryClone)
+                categoryId = cat.id
             })
             it('Should create the new offer', async () => {
                 const token = (await logInUser(user)).accessToken
                 const res = await request.post('/api/offers/').set('Authorization', `bearer ${token}`)
-                .send(createOfferClone).expect(200)
-                //logger.info('Response => ' + JSON.stringify(res.body))
+                    .send(createOfferClone).expect(200)
```
Comment (CamiloNicolasMartinez) [inline test/server/controllers/offers.test.ts:106]: We should expect something to happen. We should validate the created offer and verify that does match the provided data

**Review by CamiloNicolasMartinez — APPROVED** [review]
